crates/bevy_asset/src/server/info.rs

```diff
@@ -162,15 +163,15 @@ impl AssetInfos {
         );
         // it is ok to unwrap because TypeId was specified above
         let (handle, should_load) =
-            unwrap_with_context(result, core::any::type_name::<A>()).unwrap();
+            unwrap_with_context(result, Either::Left(core::any::type_name::<A>())).unwrap();
```

Comment:
I'm _personally_ not a fan of `Either`. I'd rather just define a little `enum` with descriptive names (rather than just `Left` and `Right`), and take the opportunity to implement `From<...>` for each of the possibilities, since we _know_ what each type option is (and thus won't need specialisation).

```rust
pub(crate) enum MyTypeInfo {
    Name(&'static str),
    TypeId(TypeId),
}

impl From<&'static str> for MyTypeInfo { ... }
impl From<TypeId> for MyTypeInfo { ... }

pub(crate) fn unwrap_with_context<T>(
    result: Result<T, GetOrCreateHandleInternalError>,
    type_info: impl Into<MyTypeInfo>,
) -> Option<T> { ... }

unwrap_with_context(result, core::any::type_name::<A>());
unwrap_with_context(result, core::any::TypeId::of::<A>());
```

Having said that, I know not everyone agrees with me on this opinion and I _do_ think this is one of the best examples of where to appropriate use `Either`. I won't hold-up merging just for my own personal preferences!

Reply:
I agree that this could be more ergonomic, using something like
```rs
enum TypeNameOrId<'a> {
    Name(&'a str),
    Id(TypeId),
}
fn foo(type_info: impl Into<TypeNameOrId<'_>>) { .. }
```
However, I'd like to leave that for a future PR, since I bet I could find some more places where this would be a nice ergonomic change, which isn't exclusive to this PR's NestedLoader API changes.
Relevant discussion here: https://discord.com/channels/691052431525675048/749332104487108618/1290249951556665444
I'll keep it as Either for now